src/session/session_api.c

```diff
@@ -540,18 +540,27 @@ __wt_open_cursor(WT_SESSION_IMPL *session, const char *uri, WT_CURSOR *owner, co
   WT_CURSOR **cursorp)
 {
     WT_DECL_RET;
+    uint64_t hash_value;
+
+    hash_value = 0;
 
     /* We should not open other cursors when there are open history store cursors in the session. */
     WT_ASSERT(session, strcmp(uri, WT_HS_URI) == 0 || session->hs_cursor_counter == 0);
 
     /* We do not cache any subordinate tables/files cursors. */
     if (owner == NULL) {
-        if ((ret = __wt_cursor_cache_get(session, uri, NULL, cfg, cursorp)) == 0)
+        __wt_cursor_get_hash(session, uri, NULL, &hash_value);
+        if ((ret = __wt_cursor_cache_get(session, uri, hash_value, NULL, cfg, cursorp)) == 0)
             return (0);
         WT_RET_NOTFOUND_OK(ret);
     }
 
-    return (__session_open_cursor_int(session, uri, owner, NULL, cfg, cursorp));
+    ret = __session_open_cursor_int(session, uri, owner, NULL, cfg, cursorp);
+
+    if (*cursorp != NULL)
+        (*cursorp)->uri_hash = hash_value;
+
+    return (ret);
```

Comment:
To get rid of repeating code (lines 560-561 is the same as 619-620), I would move the setting of `uri_hash` to `__session_open_cursor_int`.  That is, `__session_open_cursor_int` should take an additional argument `hash_value`, and set the result cursor's `uri_hash` there.

Reply:
Fixed